src/net/mod.rs

```diff
@@ -0,0 +1,100 @@
+// need this to allow MessageEnvelope to have custom serde bounds
+#![allow(clippy::type_repetition_in_bounds)]
+
+mod client;
+mod data;
+mod server;
+
+pub use client::{MpcHandle, MpcHttpConnection as Client};
+pub use data::Command;
+#[cfg(feature = "self-signed-certs")]
+pub use server::tls_config_from_self_signed_cert;
+pub use server::{BindTarget, MpcServer};
+use std::str::FromStr;
+
+use crate::net::server::MpcServerError;
+use crate::protocol::{QueryId, RecordId};
+use async_trait::async_trait;
+use axum::body::Bytes;
+use axum::extract::{FromRequest, RequestParts};
+use axum::http::header::HeaderName;
+
+/// name of the `offset` header to use for [`RecordHeaders`]
+static OFFSET_HEADER_NAME: HeaderName = HeaderName::from_static("offset");
+/// name of the `data_size` header to use for [`RecordHeaders`]
+static DATA_SIZE_HEADER_NAME: HeaderName = HeaderName::from_static("data-size");
+
+/// Headers that are expected on requests involving a batch of records.
+/// # `offset`
+/// For any given batch, their `record_id`s must be known. The first record in the batch will have id
+/// `offset`, and subsequent records will be in-order from there.
+/// # `data_size`
+/// the batch will be transmitted as a single `Bytes` block, and the receiver will need to know how
+/// to divide up the block into individual records. `data_size` represents the number of bytes each
+/// record consists of
+pub struct RecordHeaders {
+    offset: u32,
+    data_size: u32,
+}
+
+impl RecordHeaders {
```

Comment:
Might be good to have a quick design discussion on this topic. As I understood the networking/messaging layer interface, the network layer will receive streams in discrete chunks that start and end at a some arbitrary record. so for instance, messaging layer might have 2 Channels of data to send to a different helper:

H1, step1 -> records[0..100]
H2, step2 -> records[0..100]

and it might decide to split those up into 20-record chunks. so it would pass those chunks to the network layer, which would make unique POST reqs of contiguous bytes representing:

H1, step1 -> records[0,20]
H2, step1 -> records[0,20]
H1, step1 -> records[20,40]
H2, step1 -> records[20,40]
etc...

the receiving helpers' network layers would receive those contiguous chunks, break them up into [`MessageEnvelope`s](https://github.com/private-attribution/ipa/blob/cea96f47a0d3d74e1bf4efdf03e190b70097efe9/src/helpers/fabric.rs#L16), and pass them up to their messaging layers.

It sounds like you're considering a different solution. one that would have a single POST stream all records through it?

Reply:
At the networking layer, what I would expect is that each stream is a contiguous, undifferentiated sequence of records, starting from index 0 and increasing monotonically without gaps.  Different streams will be used for records from different steps or to/from different helpers.  Though your interface will accept records in chunks, you should be able to assert that each chunk you receive for a stream is contiguous with the previous one.  You shouldn't care - except to the extent that it affects performance - how big those chunks are.